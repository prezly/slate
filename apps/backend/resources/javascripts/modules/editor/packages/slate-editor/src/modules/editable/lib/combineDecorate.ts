import type { Decorate } from '@prezly/slate-commons';

export function combineDecorate(decorateFns: Decorate[]): Decorate {
    return function ({ editor, entry }) {
        return decorateFns.flatMap((decorate) => decorate({ editor, entry }));
    };
}
