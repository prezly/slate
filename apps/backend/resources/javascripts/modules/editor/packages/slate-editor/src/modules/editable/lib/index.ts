export { combineDecorate } from './combineDecorate';
export { combineOnKeyDown } from './combineOnKeyDown';
export { combineRenderElement } from './combineRenderElement';
export { combineRenderLeaf } from './combineRenderLeaf';
export { createExtensionsDecorators } from './createExtensionsDecorators';
