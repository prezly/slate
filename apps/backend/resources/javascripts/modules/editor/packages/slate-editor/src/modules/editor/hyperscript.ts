/* eslint-disable @typescript-eslint/no-namespace */

import {
    BULLETED_LIST_NODE_TYPE,
    DIVIDER_NODE_TYPE,
    LIST_ITEM_NODE_TYPE,
    LIST_ITEM_TEXT_NODE_TYPE,
    NUMBERED_LIST_NODE_TYPE,
    PARAGRAPH_NODE_TYPE,
    QUOTE_NODE_TYPE,
} from '@prezly/slate-types';
import type { ReactNode } from 'react';
import { createHyperscript, createText } from 'slate-hyperscript';

declare global {
    namespace JSX {
        interface IntrinsicElements {
            'h-divider': {
                children?: ReactNode;
            };
            'h-li': {
                children?: ReactNode;
            };
            'h-li-text': {
                children?: ReactNode;
            };
            'h-ol': {
                children?: ReactNode;
            };
            'h-p': {
                children?: ReactNode;
            };
            'h-ul': {
                children?: ReactNode;
            };
            'h:blockquote': {
                children?: ReactNode;
            };
            'h:text': {
                children?: ReactNode;
            };
        }
    }
}

export const hyperscript = createHyperscript({
    elements: {
        'h-p': { type: PARAGRAPH_NODE_TYPE },
        'h-divider': { type: DIVIDER_NODE_TYPE },
        'h-ol': { type: NUMBERED_LIST_NODE_TYPE },
        'h-ul': { type: BULLETED_LIST_NODE_TYPE },
        'h-li': { type: LIST_ITEM_NODE_TYPE },
        'h-li-text': { type: LIST_ITEM_TEXT_NODE_TYPE },
        'h:blockquote': { type: QUOTE_NODE_TYPE },
    },
    creators: {
        'h-text': createText,
        'h:text': createText,
    },
});
