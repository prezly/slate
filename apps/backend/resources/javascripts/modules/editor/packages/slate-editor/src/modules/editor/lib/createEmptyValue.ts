import { PARAGRAPH_NODE_TYPE } from '@prezly/slate-types';

export function createEmptyValue() {
    return [
        {
            children: [{ text: '' }],
            type: PARAGRAPH_NODE_TYPE,
        },
    ];
}
