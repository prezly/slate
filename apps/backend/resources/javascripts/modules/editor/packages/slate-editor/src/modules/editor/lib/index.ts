export { createEmptyValue } from './createEmptyValue';
export { createOnCut } from './createOnCut';

export { insertDivider } from './divider';

export * from './table';

export { isEditorValueEqual } from './isEditorValueEqual';
export { useCursorInView } from './useCursorInView';
