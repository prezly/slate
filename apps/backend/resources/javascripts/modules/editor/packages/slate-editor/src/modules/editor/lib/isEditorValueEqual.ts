import { isEqual } from '@technically/lodash';
import { type SlateEditor } from '@udecode/plate-common';
import type { Element, Text, Descendant } from 'slate';

import type { ElementsEqualityCheckEditor } from '../plugins';

export function isEditorValueEqual<T extends Descendant>(
    editor: SlateEditor & ElementsEqualityCheckEditor,
    a: T[],
    b: T[],
): boolean {
    const compareNodes = cached(editor, (node: Descendant, another: Descendant): boolean => {
        const isNodeText = isText(node);
        const isAnotherText = isText(another);

        if (isNodeText && isAnotherText) {
            return isEqual(node, another); // default lodash `isEqual` logic;
        }

        if (!isNodeText && !isAnotherText) {
            const equal = editor.isElementEqual(node as Element, another as Element);
            if (typeof equal !== 'undefined') {
                if (editor.isVoid(node) || editor.isVoid(another)) {
                    // Do not compare void elements children
                    return equal;
                }
                return equal && compareLists(node.children, another.children);
            }
            const { children, ...props } = node;
            const { children: otherChildren, ...otherProps } = another;

            return isEqual(props, otherProps) && compareLists(children, otherChildren);
        }

        return false; // comparing Text & Element === always false
    });

    function compareLists(nodes: Descendant[], other: Descendant[]): boolean {
        return (
            nodes.length == other.length && nodes.every((node, i) => compareNodes(node, other[i]))
        );
    }

    return a === b || compareLists(a, b);
}

function isText(node: Descendant): node is Text {
    return 'text' in node;
}

// CACHE

const CACHE: WeakMap<
    ElementsEqualityCheckEditor,
    WeakMap<Descendant, WeakMap<Descendant, boolean>>
> = new WeakMap();

type WeakMatrix = WeakMap<
    ElementsEqualityCheckEditor,
    WeakMap<Descendant, WeakMap<Descendant, boolean>>
>;
type NodesComparator = (node: Descendant, another: Descendant) => boolean;

function cached(editor: ElementsEqualityCheckEditor, fn: NodesComparator): NodesComparator {
    return (node, another) => {
        const cached = get(CACHE, editor, node, another) ?? get(CACHE, editor, another, node);

        if (typeof cached !== 'undefined') {
            return cached;
        }

        const ret = fn(node, another);

        set(CACHE, editor, node, another, ret);
        set(CACHE, editor, another, node, ret);

        return ret;
    };
}

function get(
    matrix: WeakMatrix,
    editor: ElementsEqualityCheckEditor,
    node: Descendant,
    another: Descendant,
): boolean | undefined {
    return matrix.get(editor)?.get(node)?.get(another);
}

function set(
    matrix: WeakMatrix,
    editor: ElementsEqualityCheckEditor,
    node: Descendant,
    another: Descendant,
    value: boolean,
): boolean {
    const level2 = matrix.get(editor) ?? new WeakMap();
    const level3 = level2.get(node) ?? new WeakMap();

    level2.set(node, level3);
    level3.set(another, value);

    return value;
}
