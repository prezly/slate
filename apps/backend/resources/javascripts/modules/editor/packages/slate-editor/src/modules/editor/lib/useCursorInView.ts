import { EditorCommands } from '@prezly/slate-commons';
import { isImageNode } from '@prezly/slate-types';
import { isExpanded, type SlateEditor } from '@udecode/plate-common';
import { toDOMNode, toDOMRange } from '@udecode/plate-common/react';
import jsonStableStringify from 'json-stable-stringify';
import { useLayoutEffect } from 'react';

import { ensureElementInView, ensureRangeInView } from '#lib';

export interface Parameters {
    minTop?: number;
    minBottom?: number;
}

export function useCursorInView(editor: SlateEditor, parameters: false | Parameters = false): void {
    useLayoutEffect(() => {
        if (!parameters) return;

        try {
            ensureCursorInView(editor, parameters);
        } catch {
            // Ignore all cursor-related errors. They sometimes come from Slate.
        }
    }, [editor, editor.selection, jsonStableStringify(parameters)]);
}

function ensureCursorInView(editor: SlateEditor, parameters: Parameters): void {
    if (!editor.selection) {
        return;
    }
    const [currentNode] = EditorCommands.getCurrentNodeEntry(editor) || [];

    if (isExpanded(editor.selection) || isImageNode(currentNode)) {
        // Slate has built-in mechanism to follow the cursor, but it's not perfect,
        // see: https://github.com/ianstormtaylor/slate/issues/3750
        // We don't know any issues when selecting things, so our fix is only
        // about following the cursor (when `Range.isCollapsed(editor.selection)`).
        // When `Range.isExpanded(editor.selection)` we leave the work to Slate.
        return;
    }

    if (
        currentNode &&
        EditorCommands.isBlock(editor, currentNode) &&
        EditorCommands.isVoid(editor, currentNode)
    ) {
        /**
         * Slate reports invalid `domRange` on void elements. The reported range points to
         * the `data-slate-zero-width` element which is inside [data-slate-spacer="true"]
         * element after a void block.
         * Looking at how browser renders the element, it appears as if it exists in the
         * paragraph after the void element. Therefore, positioning calculation is incorrect,
         * which causes the an edge case where scrolling does not work correctly:
         * https://github.com/prezly/prezly/pull/8434#discussion_r485427970
         * This seems to be a problem in Chrome/Chromium, but not in Firefox.
         *
         * As a fallback, we use the DOM element of the currentNode and ensure this
         * element is in view.
         */
        const domElement = toDOMNode(editor, currentNode);
        ensureElementInView(domElement, {
            minBottom: parameters.minBottom,
            minTop: parameters.minTop,
        });
        return;
    }

    const domRange = toDOMRange(editor, editor.selection);
    ensureRangeInView(domRange, {
        minBottom: parameters.minBottom,
        minTop: parameters.minTop,
        skipWhenDoesNotFitView: true,
    });
}
