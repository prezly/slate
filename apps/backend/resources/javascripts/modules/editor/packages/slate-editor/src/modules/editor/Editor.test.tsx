/** @jsx hyperscript */

import { EditorCommands } from '@prezly/slate-commons';
import fs from 'fs';
import path from 'path';
import { Editor } from 'slate';

import { createDataTransfer } from '#lib';

import { hyperscript } from './hyperscript';
import { insertDivider } from './lib';
import { createEditor } from './test-utils';

function readTestFile(filepath: string): string {
    return fs.readFileSync(path.resolve(__dirname, '__tests__', filepath), 'utf-8');
}

describe('Editor', () => {
    describe('deleteForward()', () => {
        it('should focus the list after the paragraph when using deleteForward (Del key) ', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                    <h-ul>
                        <h-li>
                            <h-li-text>
                                <h-text>lorem ipsum</h-text>
                            </h-li-text>
                        </h-li>
                    </h-ul>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-ul>
                        <h-li>
                            <h-li-text>
                                <h-text>
                                    <cursor />
                                    lorem ipsum
                                </h-text>
                            </h-li-text>
                        </h-li>
                    </h-ul>
                </editor>
            ) as unknown as Editor;

            editor.deleteForward('character'); // equivalent to pressing delete (fn+backspace on macs)

            expect(() => {
                // In other extensions we modify the behavior of `deleteForward`, which in the past
                // caused the editor to be left with an invalid selection (pointing to non-leaf nodes).
                // Calling `Editor.marks` afterwards would throw an error:
                // "Cannot get the leaf node at path [${path}] because it refers to a non-leaf node: ${node}"
                // see: https://app.clubhouse.io/prezly/story/19544/editor-crashes-when-pressing-del-in-an-empty-paragraph-before-a-list
                Editor.marks(editor);
            }).not.toThrow();

            expect(editor.children).toEqual(expected.children);
            expect(editor.selection).toEqual(expected.selection);
        });
    });

    describe('Insertion', () => {
        it('should insert block node after a list node (and not within the list)', () => {
            const editor = createEditor(
                <editor>
                    <h-ul>
                        <h-li>
                            <h-li-text>
                                <h-text>lorem ipsum</h-text>
                            </h-li-text>
                        </h-li>
                    </h-ul>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-ul>
                        <h-li>
                            <h-li-text>
                                <h-text>lorem ipsum</h-text>
                            </h-li-text>
                        </h-li>
                    </h-ul>
                    <h-divider>
                        <h-text />
                    </h-divider>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>
            ) as unknown as Editor;

            insertDivider(editor);

            expect(editor.children).toEqual(expected.children);
            expect(editor.selection).toEqual(expected.selection);
        });
    });

    describe('Pasting', () => {
        it('should deserialize Google Docs reference document', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/google-docs-reference-document.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/google-docs-reference-document.html'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        it('should deserialize Microsoft Word reference document', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/docx-reference-document.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/docx-reference-document.html'),
                    'text/rtf': readTestFile('input/docx-reference-document.rtf'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        it('should not merge sibling divs into p', function () {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/divs.html'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchSnapshot();
        });

        it('should deserialize paragraph nested in quote', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/paragraph-in-quote.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/paragraph-in-quote.html'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toEqual(JSON.parse(expected));
        });

        it('should deserialize image nested in paragraph', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/image-in-paragraph.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/image-in-paragraph.html'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        it('should deserialize image nested in h1', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/image-in-h1.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/image-in-h1.html'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        it('should deserialize list without throwing an error (normalizations count limit)', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/list-normalization-1.html'),
                },
            });

            expect(() => {
                editor.insertData(dataTransfer);
            }).not.toThrow();
        });

        it('should deserialize custom Slack line-breaks', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/slack-line-breaks.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/slack-line-breaks.html'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        it('should insert nodes with list nested in a paragraph', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const input = readTestFile('input/list-normalization-2.json');
            const expected = readTestFile('expected/list-normalization-2.json');

            EditorCommands.insertNodes(editor, JSON.parse(input));

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        /**
         * @see CARE-1320
         */
        it('should handle pasting nodes with empty children array', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const input = readTestFile('input/list-normalization-3.json');
            const expected = readTestFile('expected/list-normalization-3.json');

            EditorCommands.insertNodes(editor, JSON.parse(input), { mode: 'highest' });

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        /**
         * @see CARE-1965
         */
        it('should normalize list-items directly nested into another list-item', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const input = readTestFile('input/list-normalization-4.json');
            const expected = readTestFile('expected/list-normalization-4.json');

            editor.children = JSON.parse(input).children;

            Editor.normalize(editor, { force: true });

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        /**
         * @see CARE-1320
         */
        it('should handle nodes with empty children array', () => {
            const editor = createEditor(
                <editor>
                    <h-ol>
                        <h-li></h-li>
                    </h-ol>
                </editor>,
            );

            Editor.normalize(editor, { force: true });

            expect(editor.children).toMatchObject([
                {
                    type: 'numbered-list',
                    children: [
                        {
                            type: 'list-item',
                            children: [
                                {
                                    type: 'list-item-text',
                                    children: [{ text: '' }],
                                },
                            ],
                        },
                    ],
                },
            ]);
        });

        it('should deserialize marks from style attributes', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/marks-in-style.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/marks-in-style.html'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        it('should deserialize plaintext copied from PDF', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            <cursor />
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = readTestFile('expected/pdf-plaintext.json');
            const dataTransfer = createDataTransfer({
                items: {
                    'text/html': readTestFile('input/pdf-plaintext.html'),
                    'text/plain': readTestFile('input/pdf-plaintext.txt'),
                },
            });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchObject(JSON.parse(expected));
        });

        it('should preserve target block formatting when pasting a single text block', () => {
            const HTML = '<h1>Hello world</h1>';

            const editor = createEditor(
                <editor>
                    <h:blockquote>
                        <h:text>
                            <cursor />
                        </h:text>
                    </h:blockquote>
                </editor>,
            );

            const dataTransfer = createDataTransfer({ items: { 'text/html': HTML } });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchSnapshot();
        });

        it('should preserve source blocks formatting when pasting multiple text blocks', () => {
            const HTML = `
                 <h1>Hello world</h1>
                 <h2>This is it, a subtitle.</h2>
                 <p>How do you like it?</p>
            `;

            const editor = createEditor(
                <editor>
                    <h:blockquote>
                        <h:text>
                            <cursor />
                        </h:text>
                    </h:blockquote>
                </editor>,
            );

            const dataTransfer = createDataTransfer({ items: { 'text/html': HTML } });

            editor.insertData(dataTransfer);

            expect(editor.children).toMatchSnapshot();
        });
    });

    describe('withRootElements', () => {
        it('should lift up nested paragraphs', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                        <h-p>
                            <h-text>nested paragraph - text before</h-text>
                            <h-p>
                                <h-text>deeply nested paragraph</h-text>
                            </h-p>
                            <h-text>nested paragraph - text after</h-text>
                        </h-p>
                        <h-text>text after</h-text>
                    </h-p>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                    </h-p>
                    <h-p>
                        <h-text>nested paragraph - text before</h-text>
                    </h-p>
                    <h-p>
                        <h-text>deeply nested paragraph</h-text>
                    </h-p>
                    <h-p>
                        <h-text>nested paragraph - text after</h-text>
                    </h-p>
                    <h-p>
                        <h-text>text after</h-text>
                    </h-p>
                </editor>
            ) as unknown as Editor;

            Editor.normalize(editor, { force: true });

            expect(editor.children).toEqual(expected.children);
        });
    });
    //
    describe('Voids', () => {
        it('should properly remove text from trailing paragraph', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                    </h-p>
                    <h-divider>
                        <h-text>
                            <anchor />
                        </h-text>
                    </h-divider>
                    <h-p>
                        <h-text>
                            text <focus />
                            after
                        </h-text>
                    </h-p>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                    </h-p>
                    <h-p>
                        <h-text>
                            <cursor />
                            after
                        </h-text>
                    </h-p>
                </editor>
            ) as unknown as Editor;

            editor.deleteFragment();

            expect(editor.children).toEqual(expected.children);
            expect(editor.selection).toEqual(expected.selection);
        });

        it('should properly remove text from leading paragraph', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>
                            text <anchor />
                            before
                        </h-text>
                    </h-p>
                    <h-divider>
                        <h-text>
                            <focus />
                        </h-text>
                    </h-divider>
                    <h-p>
                        <h-text>text after</h-text>
                    </h-p>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-p>
                        <h-text>
                            text <cursor />
                        </h-text>
                    </h-p>
                    <h-p>
                        <h-text>text after</h-text>
                    </h-p>
                </editor>
            ) as unknown as Editor;

            editor.deleteFragment();

            expect(editor.children).toEqual(expected.children);
            expect(editor.selection).toEqual(expected.selection);
        });

        it('should properly remove 2 consecutive void blocks after a paragraph', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                    </h-p>
                    <h-divider>
                        <h-text>
                            <anchor />
                        </h-text>
                    </h-divider>
                    <h-divider>
                        <h-text>
                            <focus />
                        </h-text>
                    </h-divider>
                    <h-p>
                        <h-text>text after</h-text>
                    </h-p>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                    </h-p>
                    <h-p>
                        <h-text>
                            <cursor />
                            text after
                        </h-text>
                    </h-p>
                </editor>
            ) as unknown as Editor;

            editor.deleteFragment();

            expect(editor.children).toEqual(expected.children);
            expect(editor.selection).toEqual(expected.selection);
        });

        it('should properly remove 2 consecutive void blocks after 2 paragraphs', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>text before 1</h-text>
                    </h-p>
                    <h-p>
                        <h-text>text before 2</h-text>
                    </h-p>
                    <h-divider>
                        <h-text>
                            <anchor />
                        </h-text>
                    </h-divider>
                    <h-divider>
                        <h-text>
                            <focus />
                        </h-text>
                    </h-divider>
                    <h-p>
                        <h-text>text after</h-text>
                    </h-p>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-p>
                        <h-text>text before 1</h-text>
                    </h-p>
                    <h-p>
                        <h-text>text before 2</h-text>
                    </h-p>
                    <h-p>
                        <h-text>
                            <cursor />
                            text after
                        </h-text>
                    </h-p>
                </editor>
            ) as unknown as Editor;

            editor.deleteFragment();

            expect(editor.children).toEqual(expected.children);
            expect(editor.selection).toEqual(expected.selection);
        });

        it('should properly remove 4 consecutive void blocks', () => {
            const editor = createEditor(
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                    </h-p>
                    <h-divider>
                        <h-text>
                            <anchor />
                        </h-text>
                    </h-divider>
                    <h-divider>
                        <h-text />
                    </h-divider>
                    <h-divider>
                        <h-text />
                    </h-divider>
                    <h-divider>
                        <h-text>
                            <focus />
                        </h-text>
                    </h-divider>
                    <h-p>
                        <h-text>text after</h-text>
                    </h-p>
                </editor>,
            );

            const expected = (
                <editor>
                    <h-p>
                        <h-text>text before</h-text>
                    </h-p>
                    <h-p>
                        <h-text>
                            <cursor />
                            text after
                        </h-text>
                    </h-p>
                </editor>
            ) as unknown as Editor;

            editor.deleteFragment();

            expect(editor.children).toEqual(expected.children);
            expect(editor.selection).toEqual(expected.selection);
        });
    });
});
