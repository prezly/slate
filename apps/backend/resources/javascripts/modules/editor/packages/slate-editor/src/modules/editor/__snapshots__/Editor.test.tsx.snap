// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor Pasting should not merge sibling divs into p 1`] = `
[
  {
    "children": [
      {
        "text": "Bonjour Sylvain, ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "J'espère que vous allez bien ?",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Je me permets de vous contacter car je souhaitais vous proposer de vous entretenir avec ",
      },
      {
        "children": [
          {
            "bold": true,
            "text": "Bernhard Ney",
          },
        ],
        "href": "https://www.linkedin.com/in/bernhardney/",
        "new_tab": true,
        "type": "link",
      },
      {
        "bold": true,
        "text": ", Directeur International de ",
      },
      {
        "children": [
          {
            "bold": true,
            "text": "COMATCH",
          },
        ],
        "href": "https://www.comatch.com/fr/",
        "new_tab": true,
        "type": "link",
      },
      {
        "bold": true,
        "text": ", premier réseau de consultants et experts indépendants en Europe. ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "A la veille du reconfinement, je vous propose de faire le point avec lui sur l'industrie du conseil et le métier de consultant, qui passe traditionnellement beaucoup de temps en déplacement pour rencontrer ses clients et réaliser ses missions. Voici quelques sujets que vous pourriez aborder avec Bernhard:",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "children": [
          {
            "children": [
              {
                "text": "Quel est le rôle des plateformes d'intermédiation dans ce contexte ?",
              },
            ],
            "type": "list-item-text",
          },
        ],
        "type": "list-item",
      },
      {
        "children": [
          {
            "children": [
              {
                "text": "Quels impacts sur les activités de conseil et l'industrie du conseil indépendant?  ",
              },
            ],
            "type": "list-item-text",
          },
        ],
        "type": "list-item",
      },
      {
        "children": [
          {
            "children": [
              {
                "text": "",
              },
            ],
            "type": "list-item-text",
          },
        ],
        "type": "list-item",
      },
      {
        "children": [
          {
            "children": [
              {
                "text": "Comment les entreprises et les consultants se sont-ils adaptés pour poursuivre les missions?  ",
              },
            ],
            "type": "list-item-text",
          },
        ],
        "type": "list-item",
      },
      {
        "children": [
          {
            "children": [
              {
                "text": "Comment le conseil la consommation de conseil va-t'elle évoluer à l'avenir ? Plus de flexibilité, plus d'indépendants, les cabinets de conseil seront-ils toujours compétitifs?",
              },
            ],
            "type": "list-item-text",
          },
        ],
        "type": "list-item",
      },
      {
        "children": [
          {
            "children": [
              {
                "text": "Comment bien travailler avec un consultant à distance ? ",
              },
            ],
            "type": "list-item-text",
          },
        ],
        "type": "list-item",
      },
    ],
    "type": "bulleted-list",
  },
  {
    "children": [
      {
        "text": "A cette occasion, il pourra également dévoiler les résultats de l'étude "",
      },
      {
        "italic": true,
        "text": "L’ADN du consultant indépendant en 2020 : Résilience et flexibilité en temps de crise"",
      },
      {
        "text": " (enquête réalisé avant et après le début de la crise), sur les limites et tendances du conseil indépendant en cette période d'incertitudes économiques. ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "bold": true,
        "text": "Ce pourrait également être l'occasion de faire le point sur COMATCH, son modèle et sa présence en France. ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "bold": true,
        "text": "COMATCH en bref",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Le rôle de COMATCH est de mettre les entreprises - de la PME au grand groupes - en relation avec les profils adaptés à leurs besoins. Chacun des 10,000 profils internationaux inscrits sur le réseau a suivi un processus de sélection rigoureux pour assurer le meilleur service aux clients. Aujourd'hui le réseau enregistre plus de 10000 consultants principalement d'Europe et des Etats-Unis, mais également d'Asie ou encore d'Amérique du Sud. ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "COMATCH répond ainsi à la quête de réactivité et de flexibilité des entreprises dans leur recherche en compétences externes et notamment dans le contexte actuel. ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": " ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Est-ce que cette proposition vous intéresse ?",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": " ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Dans l'attente de votre retour, je vous souhaite une excellente fin de journée.",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": " ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Bien cordialement,",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": " ",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "Noémie Diaz de Cerio",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`Editor Pasting should preserve source blocks formatting when pasting multiple text blocks 1`] = `
[
  {
    "children": [
      {
        "text": "Hello world",
      },
    ],
    "type": "heading-one",
  },
  {
    "children": [
      {
        "text": "This is it, a subtitle.",
      },
    ],
    "type": "heading-two",
  },
  {
    "children": [
      {
        "text": "How do you like it?",
      },
    ],
    "type": "paragraph",
  },
  {
    "children": [
      {
        "text": "",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`Editor Pasting should preserve target block formatting when pasting a single text block 1`] = `
[
  {
    "children": [
      {
        "text": "Hello world",
      },
    ],
    "type": "block-quote",
  },
]
`;
